File: dfs/getLonelyNodes.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    List<Integer> res = new ArrayList<Integer>();
    public List<Integer> getLonelyNodes(TreeNode root) {
        dfs(root);
        return res;
    }
    public void dfs(TreeNode root) {
        if(root != null) {
            dfs(root.left);
            if(root.left != null && root.right == null) {
                res.add(root.left.val);
            }
            if(root.left == null && root.right != null) {
                res.add(root.right.val);
            }
            dfs(root.right);
        }
    }
}

File: dfs/getLonelyNodes.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def getLonelyNodes(self, root: TreeNode) -> List[int]:
        res = []
        
        def dfs(root):
            if root:
                dfs(root.left)
                if root.left is not None and root.right is None:
                    res.append(root.left.val)
                if root.left is None and root.right is not None:
                    res.append(root.right.val)
                dfs(root.right)
        dfs(root)
        return res